src/adaptive_resortable_list.h

```diff
@@ -0,0 +1,141 @@
+                        // i.e. the number of keywords found and expected
+
+    size_t relinkings;  // the number of relinkings we have made so far
+    size_t allocated;   // the number of keywords allocated
+
+    size_t rechecks;    // the number of iterations between re-checks of the
+                        // wanted number of keywords
+                        // this is only needed in cases where the source
+                        // is having less lines over time.
+
+    size_t added;       // it is non-zero if new keywords have been added
+                        // this is only needed to detect new lines have
+                        // been added to the file, over time.
+
+    // the processor to do the job
+    void (*processor)(const char *name, uint32_t hash, const char *value, void *dst);
+
+    // the linked list of the keywords
+    ARL_ENTRY *head;
+
+    // since we keep the list of keywords sorted (as found in the source data)
+    // this is next keyword that we expect to find in the source data.
+    ARL_ENTRY *next_keyword;
+} ARL_BASE;
+
+// create a new ARL
+extern ARL_BASE *arl_create(void (*processor)(const char *, uint32_t, const char *, void *), size_t rechecks);
+
+// free an ARL
+extern void arl_free(ARL_BASE *arl_base);
+
+// register an expected keyword to the ARL
+// together with its destination ( i.e. the output of the processor() )
+extern ARL_ENTRY *arl_expect(ARL_BASE *base, const char *keyword, void *dst);
+
+// an internal call to complete the check() call
+extern int arl_find_or_create_and_relink(ARL_BASE *base, const char *s, uint32_t hash, const char *value);
+
+// begin an ARL iteration
+extern void arl_begin(ARL_BASE *base);
+
+// check a keyword against the ARL
+// this is to be called for each keyword read from source data
+// s = the keyword, as collected
+// src = the src data to be passed to the processor
+// it is defined in the header file in order to be inlined
+static inline int arl_check(ARL_BASE *base, const char *keyword, const char *value) {
+    ARL_ENTRY *e = base->next_keyword;
+    uint32_t hash = simple_hash(keyword);
+
+    // it should be the first entry (pointed by base->next_keyword)
+    if(likely(hash == e->hash && !strcmp(keyword, e->name))) {
+        // it is
+
+        e->flags |= ARL_ENTRY_FLAG_FOUND;
+
+        // execute the processor
+        if(unlikely(e->dst)) {
+            base->processor(e->name, e->hash, value, e->dst);
+            base->found++;
+        }
+
+        // be prepared for the next iteration
+        base->next_keyword = e->next;
+        if(unlikely(!base->next_keyword))
+            base->next_keyword = base->head;
+
+        // stop if we collected all the values for this iteration
+        if(unlikely(base->found == base->wanted))
+            return 1;
+
+        return 0;
+    }
+
+    // we read from source, a not-expected keyword
+    return arl_find_or_create_and_relink(base, keyword, hash, value);
```

Comment:
Added two stats for testing:
```
2017-01-19 20:46:13: netdata: INFO: iteration 1, expected 15, wanted 15, allocated 92, relinkings 13, found 15, added 77, checks 92, misses 90
2017-01-19 20:46:14: netdata: INFO: iteration 2, expected 15, wanted 15, allocated 92, relinkings 13, found 15, added 0, checks 92, misses 0
2017-01-19 20:46:15: netdata: INFO: iteration 3, expected 15, wanted 15, allocated 92, relinkings 13, found 15, added 0, checks 92, misses 0
2017-01-19 20:46:16: netdata: INFO: iteration 4, expected 15, wanted 15, allocated 92, relinkings 13, found 15, added 0, checks 92, misses 0
```
"checks" is the number of times `arl_check` is called, and "misses" is the number of times it falls through to `arl_find_or_create_and_relink`.

So it only takes the slow path *once* per unexpected line. That makes this much more interesting.